dali/imgcodec/decoders/test/numpy_helper.cc

```diff
@@ -0,0 +1,235 @@
+  if (TrySkip(hdr, "True")) {
+    target.fortran_order = true;
+  } else if (TrySkip(hdr, "False")) {
+    target.fortran_order = false;
+  } else {
+    DALI_FAIL("Failed to parse fortran_order field.");
+  }
+  SkipSpaces(hdr);
+  Skip(hdr, ",");
+  SkipFieldName(hdr, "shape");
+  Skip(hdr, "(");
+  SkipSpaces(hdr);
+  while (*hdr != ')') {
+    // ParseInteger already skips the leading spaces (strtol does).
+    target.shape.shape.push_back(ParseInteger<int64_t>(hdr));
+    SkipSpaces(hdr);
+    DALI_ENFORCE(TrySkip(hdr, ",") || target.shape.size() > 1,
+                 "The first number in a tuple must be followed by a comma.");
+  }
+  if (target.fortran_order) {
+    // cheapest thing to do is to define the tensor in an reversed way
+    std::reverse(target.shape.begin(), target.shape.end());
+  }
+}
+
+void ParseHeader(FileStream *file, NumpyHeaderMeta& parsed_header) {
+  // check if the file is actually a numpy file
+  std::vector<uint8_t> token(11);
+  int64_t nread = file->Read(token.data(), 10);
+  DALI_ENFORCE(nread == 10, "Can not read header.");
+  token[nread] = '\0';
+
+  // check if heqder is too short
+  std::string header = std::string(reinterpret_cast<char*>(token.data()));
+  DALI_ENFORCE(header.find_first_of("NUMPY") != std::string::npos,
+               "File is not a numpy file.");
+
+  // extract header length
+  uint16_t header_len = 0;
+  memcpy(&header_len, &token[8], 2);
+  DALI_ENFORCE((header_len + 10) % 16 == 0,
+               "Error extracting header length.");
+
+  // read header: the offset is a magic number
+  int64 offset = 6 + 1 + 1 + 2;
+  // the header_len can be 4GiB according to the NPYv2 file format
+  // specification: https://numpy.org/neps/nep-0001-npy-format.html
+  // while this allocation could be sizable, it is performed on the host.
+  token.resize(header_len+1);
+  file->SeekRead(offset);
+  nread = file->Read(token.data(), header_len);
+  DALI_ENFORCE(nread == header_len, "Can not read header.");
+  token[header_len] = '\0';
+  header = std::string(reinterpret_cast<const char*>(token.data()));
+  DALI_ENFORCE(header.find('{') != std::string::npos, "Header is corrupted.");
+  offset += header_len;
+  file->SeekRead(offset);  // michalz: Why isn't it done when actually reading the payload?
+
+  ParseHeaderMetadata(parsed_header, header);
+  parsed_header.data_offset = offset;
+}
+
+static void TransposeHelper(SampleView<CPUBackend> output, ConstSampleView<CPUBackend> input) {
+  int n_dims = input.shape().sample_dim();
+  SmallVector<int, 6> perm;
+  perm.resize(n_dims);
+  for (int i = 0; i < n_dims; ++i)
+    perm[i] = n_dims - i - 1;
+  TYPE_SWITCH(input.type(), type2id, T, NUMPY_ALLOWED_TYPES, (
+    kernels::TransposeGrouped(view<T>(output), view<const T>(input), make_cspan(perm));
+  ), DALI_FAIL(make_string("Unsupported input type: ", input.type())));  // NOLINT
+}
+
+Tensor<CPUBackend> ReadNumpy(const std::string &path) {
+  auto file = FileStream::Open(path, false, false);
```

Comment:
I'd suggest splitting this function in two:
```suggestion
Tensor<CPUBackend> ReadNumpy(const std::string &path) {
  auto file = FileStream::Open(path, false, false);
  ReadNumpy(file);
}

Tensor<CPUBackend> ReadNumpy(InputStream &stream) {
```